rare/components/tabs/games/game_widgets/__init__.py

```diff
@@ -0,0 +1,216 @@
+from typing import Tuple, List, Union
+
+from PyQt5.QtCore import QObject, pyqtSlot
+from PyQt5.QtWidgets import QWidget
+from legendary.core import LegendaryCore
+
+from rare.models.game import RareGame
+from rare.models.signals import GlobalSignals
+from rare.shared import LegendaryCoreSingleton, GlobalSignalsSingleton
+from .icon_game_widget import IconGameWidget
+from .list_game_widget import ListGameWidget
+
+
+class LibraryWidgetController(QObject):
+    _icon_container: QWidget
+    _list_container: QWidget
+    _game_utils = None
+
+    def __init__(
+        self, icon_container: QWidget, list_container: QWidget, game_utils, parent: QWidget = None
+    ):
+        super(LibraryWidgetController, self).__init__(parent=parent)
+        self._icon_container = icon_container
+        self._list_container = list_container
+        self._game_utils = game_utils
+        self.core: LegendaryCore = LegendaryCoreSingleton()
+        self.signals: GlobalSignals = GlobalSignalsSingleton()
+
+        self.signals.progress.started.connect(self.start_progress)
+        self.signals.progress.value.connect(self.update_progress)
+        self.signals.progress.finished.connect(self.finish_progress)
+
+        self.signals.game.installed.connect(self.on_install)
+        self.signals.game.uninstalled.connect(self.on_uninstall)
+        self.signals.game.verified.connect(self.on_verified)
+
+    def add_game(self, rgame: RareGame, utils):
+        return self.add_widgets(rgame, utils)
+
+    def add_widgets(self, rgame: RareGame, utils) -> Tuple[IconGameWidget, ListGameWidget]:
+        icon_widget = IconGameWidget(rgame, utils)
+        list_widget = ListGameWidget(rgame, utils)
+        return icon_widget, list_widget
+
+    def __find_widget(
+        self, app_name: str
+    ) -> Tuple[Union[IconGameWidget, None], Union[ListGameWidget, None]]:
+        iw = self._icon_container.findChild(IconGameWidget, app_name)
+        lw = self._list_container.findChild(ListGameWidget, app_name)
+        return iw, lw
+
+    @staticmethod
+    def __visibility(widget, filter_name, search_text):
+        if filter_name == "installed":
+            visible = widget.rgame.is_installed
+        elif filter_name == "offline":
+            visible = widget.rgame.is_offline
+        elif filter_name == "32bit":
+            visible = widget.rgame.is_win32
+        elif filter_name == "mac":
+            visible = widget.rgame.is_mac
+        elif filter_name == "installable":
+            visible = not widget.rgame.is_non_asset
+        elif filter_name == "include_ue":
+            visible = True
+        elif filter_name == "all":
+            visible = not widget.rgame.is_unreal
+        else:
+            visible = True
```

Comment:
Maybe this could be prettier through a dict which maps filter name to visibility. But the logic is something we have to walk through anyways.

Reply:
Not a bad idea. But this might be clearer to read since I don't have a better way logic-wise to do it.